[SPIRV][HLSL] Add Sema and CodeGen for implicit typed buffer counters

s-perron opened the pull request:
This commit implements the Sema and CodeGen portions of the typed buffer
counter proposal described in the HLSL WG proposal 0023.

This change introduces the necessary Sema and CodeGen logic to handle
implicit counter variables for typed buffers. This includes:

- Extending `HLSLResourceBindingAttr` to store the implicit counter
  binding order ID.
- Introducing the `__builtin_hlsl_resource_counterhandlefromimplicitbinding`
  builtin.
- Updating `SemaHLSL` to correctly initialize global resource declarations
  and resource arrays with implicit counter buffers.
- Adding CodeGen support for the new builtin, which generates a
  `llvm.spv.resource.counterhandlefromimplicitbinding` intrinsic for the
  SPIR-V target and aliases the main resource handle for the DXIL target.
- Adding and updating tests to verify the new functionality.

Fixes #137032

s-perron committed:
[SPIRV][HLSL] Add Sema and CodeGen for implicit typed buffer counters

This commit implements the Sema and CodeGen portions of the typed buffer
counter proposal described in the HLSL WG proposal 0023.

This change introduces the necessary Sema and CodeGen logic to handle
implicit counter variables for typed buffers. This includes:

- Extending `HLSLResourceBindingAttr` to store the implicit counter
  binding order ID.
- Introducing the `__builtin_hlsl_resource_counterhandlefromimplicitbinding`
  builtin.
- Updating `SemaHLSL` to correctly initialize global resource declarations
  and resource arrays with implicit counter buffers.
- Adding CodeGen support for the new builtin, which generates a
  `llvm.spv.resource.counterhandlefromimplicitbinding` intrinsic for the
  SPIR-V target and aliases the main resource handle for the DXIL target.
- Adding and updating tests to verify the new functionality.

Fixes #137032

s-perron commented:
@hekota This depends on #161752. When that is merged, I will rebase this PR, and merge it.

s-perron commented:
@luciechoi Can you take a look at this PR?

luciechoi commented:
LGTM

hekota commented on clang/test/CodeGenHLSL/resources/StructuredBuffers-constructors.hlsl:
```diff
@@ -24,56 +23,60 @@ export void foo() {
-
-// Buf3 initialization part 2 - body of AppendStructuredBuffer<float> default C1 constructor that calls
-// the default C2 constructor
 // CHECK: define linkonce_odr hidden void @hlsl::StructuredBuffer<float>::StructuredBuffer()(ptr {{.*}} %this)
```
`// Buf3 initialization part 1`

Has this been removed intentionally? It should probably stay in, although the // CHECK: following it seems to be incorrectly refering to `@hlsl::StructuredBuffer<float>::StructuredBuffer()`.

hekota reviewed (commented):
Glad to see this taking shape! Please update the PR title :)

s-perron committed:
Refactor: Address review comments for implicit typed buffer counters

This commit addresses review comments on the initial implementation of implicit typed buffer counters. The changes include:

- Refactored `CGHLSLBuiltins.cpp` to move `HandleTy` declaration into the SPIR-V specific block and add an early exit for the DXIL case.
- Updated `CGHLSLRuntime.cpp` to use a `const char *` for `CreateMethodName` to avoid string allocation.
- In `HLSLBuiltinTypeDeclBuilder.cpp`, replaced `if (Record->isCompleteDefinition())` with an assert and added descriptive comments to `addCreateFromBindingWithImplicitCounter` and `addCreateFromImplicitBindingWithImplicitCounter`.
- In `StructuredBuffers-constructors.hlsl`, updated the alignment check to be more specific and re-inserted a removed comment block.
- In `SemaHLSL.cpp`, changed `CreateMethodName` to `const char *`, removed a debug print, moved the re-creation of the binding object into the `else` block, and used `AST.getHLSLAttributedResourceType` for consistency.

hekota reviewed (approved):
LGTM!
